csharp/src/Drivers/Apache/Hive2/HiveServer2Reader.cs

```diff
@@ -79,14 +81,34 @@ public HiveServer2Reader(
 
         public async ValueTask<RecordBatch?> ReadNextRecordBatchAsync(CancellationToken cancellationToken = default)
         {
+            // All records have been exhausted
             if (_statement == null)
             {
                 return null;
             }
 
-            var request = new TFetchResultsReq(_statement.OperationHandle, TFetchOrientation.FETCH_NEXT, _statement.BatchSize);
-            TFetchResultsResp response = await _statement.Connection.Client.FetchResults(request, cancellationToken);
+            // This is the initial case, before the first fetch
+            _fetchResultResponseTask ??= FetchNext(_statement, cancellationToken);
 
+            // Await the fetch response
+            TFetchResultsResp response = await _fetchResultResponseTask;
+
+            // Build the current batch
+            RecordBatch result = CreateBatch(response, out int fetchedRows);
+
+            if ((_statement.BatchSize > 0 && fetchedRows < _statement.BatchSize) || fetchedRows == 0)
+                // This is the last batch
+                _statement = null;
```

Comment:
Can you put braces around the bodies of both the if and the else clause? They're obviously not required, but in my experience once the body becomes multi-line, it's a lot easier to make a mistake in the future unless the braces are present.

Reply:
Should the fetch task be set to null too once we're done fetching?

Reply:
Done.